Mathlib/Probability/Distributions/Uniform.lean

```diff
@@ -28,35 +35,76 @@ This file defines a number of uniform `PMF` distributions from various inputs,
 
 `PMF.ofMultiset` draws randomly from the given `Multiset`, treating duplicate values as distinct.
   Each probability is given by the count of the element divided by the size of the `Multiset`
+
+# To Do:
+* Refactor the `PMF` definitions to come from a `uniformMeasure` on a `Finset`/`Fintype`/`Multiset`.
 -/
 
 open scoped Classical MeasureTheory BigOperators NNReal ENNReal
 
-open TopologicalSpace MeasureTheory.Measure
+open TopologicalSpace MeasureTheory.Measure PMF
+
+noncomputable section
 
 namespace MeasureTheory
 
 variable {E : Type*} [MeasurableSpace E] {m : Measure E} {μ : Measure E}
-variable {Ω : Type*} {_ : MeasurableSpace Ω} {ℙ : Measure Ω}
 
+/-- A measure is a uniform measure for a set `s` if it is the rescaled restriction of the measure to
+this set.  -/
+def uniformMeasure (s : Set E) (μ : Measure E := by volume_tac) : Measure E :=
+  (μ s)⁻¹ • μ.restrict s
 
-section
+namespace UniformMeasure
 
-/-! **Uniform Distribution based on a measure** -/
+theorem absolutelyContinuous {s : Set E} :
+    uniformMeasure s μ ≪ μ := by
+  intro t ht
+  unfold uniformMeasure
+  rw [smul_apply, smul_eq_mul]
+  apply mul_eq_zero.mpr
+  refine Or.inr (le_antisymm ?_ (zero_le _))
+  exact ht ▸ restrict_apply_le s t
+
+theorem uniformMeasure_apply {s : Set E} {A : Set E}
+    (hA : MeasurableSet A) : uniformMeasure s μ A = μ (s ∩ A) / μ s := by
+  unfold uniformMeasure
+  rw [smul_apply, restrict_apply hA, ENNReal.div_eq_inv_mul, smul_eq_mul, Set.inter_comm]
+
+theorem isProbabilityMeasure {s : Set E} (hns : μ s ≠ 0) (hnt : μ s ≠ ∞) :
```

Comment:
Can you add a instance saying uniformMeasure is always a finite measure?

Reply:
Added an instance for IsFiniteMeasure and IsProbabilityMeasure. I'd guess the later is more informative, so we might prefer to have that one, as the other one should be automatically inferable from that one? I'm not sure about this specific case though, as we have these two hypotheses?

Reply:
Build fails, it can't unify the hypotheses {hns ...} and {hnt ...}. so I've removed the instances. I used

```
instance uniformMeasure.isProbabilityMeasure {s : Set E} {hns : μ s ≠ 0} {hnt : μ s ≠ ∞} :
    IsProbabilityMeasure (uniformMeasure s μ) :=
  MeasureTheory.UniformMeasure.isProbabilityMeasure hns hnt

instance uniformMeasure.isFiniteMeasure {s : Set E} {hns : μ s ≠ 0} {hnt : μ s ≠ ∞} :
    IsFiniteMeasure (uniformMeasure s μ) := by
  apply uniformMeasure.isProbabilityMeasure.toIsFiniteMeasure
  exact hns
  exact hnt
```

Reply:
You shouldn't need `{hns : μ s ≠ 0} {hnt : μ s ≠ ∞}` for showing it is a finite measure since it is the 0 measure in that case. As a result you can make the `isFiniteMeasure` a instance but not `isProbabilityMeasure`

Reply:
@JasonKYi I tried proving it like this, but the intermediate result that I need is false, can you tell me how I should proceed? 

```
instance uniformMeasure.isFiniteMeasure {s : Set E} :
    IsFiniteMeasure (uniformMeasure s μ) where
  measure_univ_lt_top := by
    have : restrict μ s Set.univ ≤ μ s := by
      simp_all only [MeasurableSet.univ, restrict_apply, Set.univ_inter, le_refl]
    unfold uniformMeasure
    have : (μ s)⁻¹ * μ s < ⊤ := by
      sorry
    simp_all only [MeasurableSet.univ, restrict_apply, Set.univ_inter, le_refl, smul_toOuterMeasure,
      OuterMeasure.coe_smul, Pi.smul_apply, smul_eq_mul]
```

Reply:
Thank you, I've fixed it. The `have : ` statement seems to be like it should be in Mathlib, but I couldn't find it.

Reply:
fixed!